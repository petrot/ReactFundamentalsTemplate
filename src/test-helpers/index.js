export * from "./local-storage.mock";
export * from "./store.mock";
export * from "./test-wrapper.mock";
