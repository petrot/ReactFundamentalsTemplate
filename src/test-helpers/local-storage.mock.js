export const mockGetItem = jest.fn();
export const mockSetItem = jest.fn();
export const mockRemoveItem = jest.fn();

export const prepareMockLocalStorage = () => {
  Object.defineProperty(window, "localStorage", {
    value: {
      getItem: (...args) => mockGetItem(...args),
      setItem: (...args) => mockSetItem(...args),
      removeItem: (...args) => mockRemoveItem(...args),
    },
  });
};
