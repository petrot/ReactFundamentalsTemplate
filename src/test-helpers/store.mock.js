export const TEST_NAME = "Test Name";

export const mockedState = {
  user: {
    isAuth: true,
    name: TEST_NAME,
    role: "admin",
    token: "mock-token",
  },
  courses: [
    {
      title: "Course 1",
      description: "Description 1",
      duration: 5,
      authors: [
        "9b87e8b8-6ba5-40fc-a439-c4e30a373d36",
        "1c972c52-3198-4098-b6f7-799b45903199",
        "072fe3fc-e751-4745-9af5-aa9eed0ea9ed",
      ],
      creationDate: "11/03/2024",
      id: "b28fd9ad-c683-4900-bf99-38a1c5748309",
    },
  ],
  authors: [
    {
      name: "author",
      id: "9b87e8b8-6ba5-40fc-a439-c4e30a373d36",
    },
    {
      name: "author2",
      id: "1c972c52-3198-4098-b6f7-799b45903199",
    },
    {
      name: "author3",
      id: "072fe3fc-e751-4745-9af5-aa9eed0ea9ed",
    },
  ],
};

export const mockedStore = {
  getState: () => mockedState,
  subscribe: jest.fn(),
  dispatch: jest.fn(),
};
