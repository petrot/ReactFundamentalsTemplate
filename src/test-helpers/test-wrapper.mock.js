import { Provider } from "react-redux";
import { MemoryRouter as Router } from "react-router-dom";
import { mockedStore } from "./store.mock";

export const TestWrapper = ({ children, store = mockedStore }) => (
  <Provider store={store}>
    <Router>{children}</Router>
  </Provider>
);
