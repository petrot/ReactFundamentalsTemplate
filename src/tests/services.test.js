import {
  createAuthor,
  createCourse,
  createUser,
  deleteCourseService,
  getAuthors,
  getCourses,
  getCurrentUser,
  login,
  logout,
  updateCourseService,
} from "../services";
import { mockedState } from "../test-helpers";

const token = "mockToken";

describe("[Services]", () => {
  it("createUser function fetches the register endpoint correctly", async () => {
    const data = { name: "John Doe" };

    global.fetch = jest.fn(() =>
      Promise.resolve({ json: () => Promise.resolve({ response: data }) })
    );

    const response = await createUser(data);

    expect(fetch).toHaveBeenCalledWith("http://localhost:4000/register", {
      method: "POST",
      body: JSON.stringify(data),
      headers: {
        "Content-Type": "application/json",
      },
    });

    expect(response).toMatchSnapshot();
  });

  it("login function fetches the login endpoint correctly", async () => {
    const data = { email: "john@doe.com", password: "asdfg" };

    global.fetch = jest.fn(() =>
      Promise.resolve({ json: () => Promise.resolve({ response: data }) })
    );

    const response = await login(data);

    expect(fetch).toHaveBeenCalledWith("http://localhost:4000/login", {
      method: "POST",
      body: JSON.stringify(data),
      headers: {
        "Content-Type": "application/json",
      },
    });

    expect(response).toMatchSnapshot();
  });

  it("getCourses function fetches the courses/all endpoint correctly", async () => {
    global.fetch = jest.fn(() =>
      Promise.resolve({
        json: () => Promise.resolve({ response: mockedState.courses }),
      })
    );

    const response = await getCourses();

    expect(fetch).toHaveBeenCalledWith("http://localhost:4000/courses/all", {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    expect(response).toMatchSnapshot();
  });

  it("getAuthors function fetches the authors/all endpoint correctly", async () => {
    global.fetch = jest.fn(() =>
      Promise.resolve({
        json: () => Promise.resolve({ response: mockedState.authors }),
      })
    );

    const response = await getAuthors();

    expect(fetch).toHaveBeenCalledWith("http://localhost:4000/authors/all", {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    expect(response).toMatchSnapshot();
  });

  it("getCurrentUser function fetches the users/me endpoint correctly", async () => {
    global.fetch = jest.fn(() =>
      Promise.resolve({
        json: () =>
          Promise.resolve({ response: { name: "John Doe", role: "admin" } }),
      })
    );

    const response = await getCurrentUser(token);

    expect(fetch).toHaveBeenCalledWith("http://localhost:4000/users/me", {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
        Authorization: token,
      },
    });

    expect(response).toMatchSnapshot();
  });

  it("updateCourseService function fetches the courses endpoint correctly", async () => {
    const data = { ...mockedState.courses[0], title: "New Title" };
    const { id, ...dataWithoutId } = data;

    global.fetch = jest.fn(() =>
      Promise.resolve({ json: () => Promise.resolve({ response: data }) })
    );

    const response = await updateCourseService(data, token);

    expect(fetch).toHaveBeenCalledWith("http://localhost:4000/courses/" + id, {
      method: "PUT",
      body: JSON.stringify(dataWithoutId),
      headers: {
        "Content-Type": "application/json",
        Authorization: token,
      },
    });

    expect(response).toMatchSnapshot();
  });

  it("logout function fetches the logout endpoint correctly", async () => {
    global.fetch = jest.fn(() =>
      Promise.resolve({ json: () => Promise.resolve({ response: "OK" }) })
    );

    const response = await logout(token);

    expect(fetch).toHaveBeenCalledWith("http://localhost:4000/logout", {
      method: "DELETE",
      headers: {
        "Content-Type": "application/json",
        Authorization: token,
      },
    });

    expect(response).toMatchSnapshot();
  });

  it("deleteCourseService function fetches the courses endpoint correctly", async () => {
    global.fetch = jest.fn(() =>
      Promise.resolve({ json: () => Promise.resolve({ response: "OK" }) })
    );

    const response = await deleteCourseService(
      mockedState.courses[0].id,
      token
    );

    expect(fetch).toHaveBeenCalledWith(
      "http://localhost:4000/courses/" + mockedState.courses[0].id,
      {
        method: "DELETE",
        headers: {
          "Content-Type": "application/json",
          Authorization: token,
        },
      }
    );

    expect(response).toMatchSnapshot();
  });

  it("createCourse function fetches the courses/add endpoint correctly", async () => {
    const { id, ...data } = mockedState.courses[0];

    global.fetch = jest.fn(() =>
      Promise.resolve({ json: () => Promise.resolve({ response: data }) })
    );

    const response = await createCourse(data, token);

    expect(fetch).toHaveBeenCalledWith("http://localhost:4000/courses/add", {
      method: "POST",
      body: JSON.stringify(data),
      headers: {
        "Content-Type": "application/json",
        Authorization: token,
      },
    });

    expect(response).toMatchSnapshot();
  });

  it("createCourse function fetches the authors/add endpoint correctly", async () => {
    const { id, ...data } = mockedState.authors[0];

    global.fetch = jest.fn(() =>
      Promise.resolve({ json: () => Promise.resolve({ response: data }) })
    );

    const response = await createAuthor(data, token);

    expect(fetch).toHaveBeenCalledWith("http://localhost:4000/authors/add", {
      method: "POST",
      body: JSON.stringify(data),
      headers: {
        "Content-Type": "application/json",
        Authorization: token,
      },
    });

    expect(response).toMatchSnapshot();
  });
});
