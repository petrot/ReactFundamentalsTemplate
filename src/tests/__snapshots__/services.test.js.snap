// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Services] createCourse function fetches the authors/add endpoint correctly 1`] = `
Object {
  "response": Object {
    "name": "author",
  },
}
`;

exports[`[Services] createCourse function fetches the courses/add endpoint correctly 1`] = `
Object {
  "response": Object {
    "authors": Array [
      "9b87e8b8-6ba5-40fc-a439-c4e30a373d36",
      "1c972c52-3198-4098-b6f7-799b45903199",
      "072fe3fc-e751-4745-9af5-aa9eed0ea9ed",
    ],
    "creationDate": "11/03/2024",
    "description": "Description 1",
    "duration": 5,
    "title": "Course 1",
  },
}
`;

exports[`[Services] createUser function fetches the register endpoint correctly 1`] = `
Object {
  "response": Object {
    "name": "John Doe",
  },
}
`;

exports[`[Services] deleteCourseService function fetches the courses endpoint correctly 1`] = `
Object {
  "response": "OK",
}
`;

exports[`[Services] getAuthors function fetches the authors/all endpoint correctly 1`] = `
Object {
  "response": Array [
    Object {
      "id": "9b87e8b8-6ba5-40fc-a439-c4e30a373d36",
      "name": "author",
    },
    Object {
      "id": "1c972c52-3198-4098-b6f7-799b45903199",
      "name": "author2",
    },
    Object {
      "id": "072fe3fc-e751-4745-9af5-aa9eed0ea9ed",
      "name": "author3",
    },
  ],
}
`;

exports[`[Services] getCourses function fetches the courses/all endpoint correctly 1`] = `
Object {
  "response": Array [
    Object {
      "authors": Array [
        "9b87e8b8-6ba5-40fc-a439-c4e30a373d36",
        "1c972c52-3198-4098-b6f7-799b45903199",
        "072fe3fc-e751-4745-9af5-aa9eed0ea9ed",
      ],
      "creationDate": "11/03/2024",
      "description": "Description 1",
      "duration": 5,
      "id": "b28fd9ad-c683-4900-bf99-38a1c5748309",
      "title": "Course 1",
    },
  ],
}
`;

exports[`[Services] getCurrentUser function fetches the users/me endpoint correctly 1`] = `
Object {
  "response": Object {
    "name": "John Doe",
    "role": "admin",
  },
}
`;

exports[`[Services] login function fetches the login endpoint correctly 1`] = `
Object {
  "response": Object {
    "email": "john@doe.com",
    "password": "asdfg",
  },
}
`;

exports[`[Services] logout function fetches the logout endpoint correctly 1`] = `
Object {
  "json": [Function],
}
`;

exports[`[Services] updateCourseService function fetches the courses endpoint correctly 1`] = `
Object {
  "response": Object {
    "authors": Array [
      "9b87e8b8-6ba5-40fc-a439-c4e30a373d36",
      "1c972c52-3198-4098-b6f7-799b45903199",
      "072fe3fc-e751-4745-9af5-aa9eed0ea9ed",
    ],
    "creationDate": "11/03/2024",
    "description": "Description 1",
    "duration": 5,
    "id": "b28fd9ad-c683-4900-bf99-38a1c5748309",
    "title": "New Title",
  },
}
`;
